pandas/_libs/tslibs/conversion.pyx

```diff
@@ -154,45 +161,48 @@ cpdef inline (int64_t, int) precision_from_unit(str unit):
     """
     cdef:
         int64_t m
+        int64_t multiplier
         int p
         NPY_DATETIMEUNIT reso = abbrev_to_npy_unit(unit)
 
+    if out_reso == NPY_DATETIMEUNIT.NPY_FR_ns:
+        multiplier = 1_000_000_000
+    elif out_reso == NPY_DATETIMEUNIT.NPY_FR_us:
+        multiplier = 1_000_000
+    elif out_reso == NPY_DATETIMEUNIT.NPY_FR_ms:
+        multiplier = 1_000
+    elif out_reso == NPY_DATETIMEUNIT.NPY_FR_s:
+        multiplier = 1
+    else:
+        raise ValueError(f"Invalid out_reso: {out_reso}")
+
     if reso == NPY_DATETIMEUNIT.NPY_FR_Y:
         # each 400 years we have 97 leap years, for an average of 97/400=.2425
         #  extra days each year. We get 31556952 by writing
         #  3600*24*365.2425=31556952
-        m = 1_000_000_000 * 31556952
-        p = 9
+        m = multiplier * 31556952
     elif reso == NPY_DATETIMEUNIT.NPY_FR_M:
         # 2629746 comes from dividing the "Y" case by 12.
-        m = 1_000_000_000 * 2629746
-        p = 9
+        m = multiplier * 2629746
     elif reso == NPY_DATETIMEUNIT.NPY_FR_W:
-        m = 1_000_000_000 * 3600 * 24 * 7
-        p = 9
+        m = multiplier * 3600 * 24 * 7
     elif reso == NPY_DATETIMEUNIT.NPY_FR_D:
-        m = 1_000_000_000 * 3600 * 24
-        p = 9
+        m = multiplier * 3600 * 24
     elif reso == NPY_DATETIMEUNIT.NPY_FR_h:
-        m = 1_000_000_000 * 3600
-        p = 9
+        m = multiplier * 3600
     elif reso == NPY_DATETIMEUNIT.NPY_FR_m:
-        m = 1_000_000_000 * 60
-        p = 9
+        m = multiplier * 60
     elif reso == NPY_DATETIMEUNIT.NPY_FR_s:
-        m = 1_000_000_000
-        p = 9
+        m = multiplier
     elif reso == NPY_DATETIMEUNIT.NPY_FR_ms:
-        m = 1_000_000
-        p = 6
+        m = multiplier // 1_000
     elif reso == NPY_DATETIMEUNIT.NPY_FR_us:
-        m = 1000
-        p = 3
+        m = multiplier // 1_000_000
     elif reso == NPY_DATETIMEUNIT.NPY_FR_ns or reso == NPY_DATETIMEUNIT.NPY_FR_GENERIC:
-        m = 1
-        p = 0
+        m = multiplier // 1_000_000_000
     else:
         raise ValueError(f"cannot cast unit {unit}")
+    p = min(9, len(str(m))-1)
```

Comment:
the str here seems like a hack.  no good alternatives?

Reply:
how about `int(math.log10(n))`?

Reply:
i guess the str call doesn't have to go into python space, so is probably preferred

Reply:
`from libc.math cimport log10` seems to work, i'll go with that